### PR TITLE
feat: TunnelForge 마이그레이션 인텔리전스 이식

### PR DESCRIPTION
**Pull request by sanghyun-io**
## Summary

TunnelForge의 마이그레이션 인텔리전스를 mysql-upgrade-checker 웹앱에 이식하여 분석 품질과 사용자 경험을 대폭 향상시킵니다.

- **Domain Model & Infrastructure**: 공통 도메인 스키마, XSS sanitizer, 에러 계층 구조 추가
- **Rule Coverage Gap**: TunnelForge에만 있던 스키마 검증 규칙 보완 (trigger/event/spatial/JSON_TABLE/DEFINER)
- **Context-Aware Analysis**: FK 의존성 그래프 빌더 + charset cascade 분석 (순환참조, 자기참조, 누락 테이블 처리)
- **Multi-Option Smart Fix**: 이슈당 2~4개 fix 옵션 + effort/risk 평가 + 컨텍스트 기반 자동 추천
- **Migration Plan Generator**: 4-phase 실행 계획 자동 생성 (Pre-flight → Preparation → Execution → Validation) + rollback SQL
- **Pre-flight & Server Query**: Pre-flight 체크리스트 + 서버 쿼리 결과 자동 분석 (VERSION/VARIABLES/GRANTS/PROCESSLIST)
- **FK Dependency Visualization**: SVG 기반 FK 관계도 (zero-dep Sugiyama layout, 200 node cap, fallback list view)
- **Migration Readiness Report**: 0-100 준비도 점수 + 리스크 등급 + HTML/Markdown 리포트

### Key Stats
- **47 files** changed, **+7,321 lines** added
- **569 tests** (기존 334 + 신규 235), all passing
- Zero runtime dependency 원칙 유지
- 모든 렌더링 경로에 XSS sanitization + CSP 적용

## Test plan

- [x] `npx tsc --noEmit` TypeScript 타입 체크 통과
- [x] `npm run test:run` 569개 전체 테스트 통과
- [x] `npm run build` 프로덕션 빌드 성공
- [ ] `npm run preview`로 브라우저 수동 검증

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/scripts/analysis/charset-cascade.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+  for (const edge of allEdges) {
+    const childTable = tables.get(edge.childTable) ??
+      findTableCaseInsensitive(tables, edge.childTable);
+    const parentTable = tables.get(edge.parentTable) ??
+      findTableCaseInsensitive(tables, edge.parentTable);
+
+    if (!childTable || !parentTable) continue;
+
+    // Check each FK column pair
+    for (let i = 0; i < edge.childColumns.length; i++) {
```
Comment (sanghyun-io) [inline src/scripts/analysis/charset-cascade.ts:129]: ### Issue #1 [HIGH]: Charset mismatch flags non-text FK columns

INT/BIGINT FK columns inherit table charset via `resolveCharset` and get flagged as charset mismatches, creating false positives.

**Suggested fix**: Gate charset/collation checks to character/string column families only (CHAR/VARCHAR/TEXT/ENUM/SET).

File: src/scripts/analysis/charset-cascade.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+/**
+ * Check if the charset mismatch is specifically a utf8mb3 <-> utf8mb4 mismatch
+ * which triggers Error 3780 during upgrade.
+ */
+function isCharsetUpgradeMismatch(charset1: string, charset2: string): boolean {
+  const pair = [charset1, charset2].sort();
+  return (pair[0] === 'utf8mb3' && pair[1] === 'utf8mb4');
+}
+
+function generateCharsetFixSQL(
```
Comment (sanghyun-io) [inline src/scripts/analysis/charset-cascade.ts:270]: ### Issue #2 [HIGH]: Auto-fix SQL forces VARCHAR(255)

`generateCharsetFixSQL` rewrites FK columns as `VARCHAR(255)` regardless of original type/nullability/default. This is unsafe and can corrupt schema.

**Suggested fix**: Preserve original column type/length/null/default and only change CHARACTER SET/COLLATE clauses. When full metadata unavailable, emit manual template instead.

File: src/scripts/fix-options/strategies/date-fix.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ * Handles: invalid_date_zero, invalid_datetime_zero
+ */
+
+import type { FixOption } from '../../domain/fix-option';
+import type { Issue } from '../../types';
+
+export function generateDateFixOptions(issue: Issue): FixOption[] {
+  const table = issue.tableName ?? 'unknown';
+  const column = issue.columnName ?? 'unknown';
+  const isNullable = issue.columnContext?.nullable ?? true;
```
Comment (sanghyun-io) [inline src/scripts/fix-options/strategies/date-fix.ts]: ### Issue #6 [MED]: nullable defaults to true when metadata missing

`?? true` default means NOT NULL columns may get `SET col = NULL` recommendation.

**Suggested fix**: Default to `nullable = false` (conservative) unless proven otherwise.

File: src/scripts/migration-plan/plan-generator.ts [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+function generateExecutionPhase(groups: OrderedFixGroup[]): MigrationPhase {
+  const steps: PlanStep[] = [];
+  let order = 1;
+
+  for (const group of groups) {
+    if (group.tableName === '__config__') continue;
+
+    for (const issue of group.issues) {
+      const rollback = generateRollbackEntry(issue);
+      const fixOption = issue.fixOptions?.find(o => o.isRecommended) ?? issue.fixOptions?.[0];
```
Comment (sanghyun-io) [inline src/scripts/migration-plan/plan-generator.ts]: ### Issue #7 [MED]: Silently omits fixable issues

When recommended option has `sqlTemplate = null`, the issue is skipped even if other options have executable SQL.

**Suggested fix**: Fallback to first SQL-bearing option when recommended has no SQL.

**Review by sanghyun-io — COMMENTED** [review]
## PR Review (GPT-5.3 Codex)

### Summary
47 files, +7,321 lines. FK dependency graph, charset cascade analysis, multi-option fix strategies, 4-phase migration plan, SVG visualization, readiness report 추가. **2 HIGH + 8 MED + 1 LOW + 1 MED = 12 issues total.**

### Findings

| # | Severity | Area | Issue |
|---|:--------:|------|-------|
| 1 | **HIGH** | Impl | Charset mismatch logic flags non-text FK columns |
| 2 | **HIGH** | Impl | Auto-fix SQL forces VARCHAR(255) — schema corruption risk |
| 3 | MED | Arch | Public APIs leak mutable internal state |
| 4 | MED | Arch | IFKEdge missing column-level info |
| 5 | MED | Arch | Visualization drops issue tables with no FK edges |
| 6 | MED | Impl | Date fix defaults nullable=true when unknown |
| 7 | MED | Impl | Execution plan silently omits fixable issues |
| 8 | MED | Test | Robustness tests weak assertions |
| 9 | MED | Test | Charset tests encode incorrect behavior |
| 10 | LOW | Test | Performance tests wall-clock brittle |
| 11 | MED | Perf | computeExecutionOrder O(n²) |
| 12 | MED | Perf | Sugiyama layer assignment O(n²) |

### Verdict
**REJECT** — 2 blocking correctness issues (false-positive charset errors + schema-corrupting auto-fix SQL). Needs revision.

---
*Reviewed by GPT-5.3 Codex*

File: src/scripts/migration-plan/execution-order.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    // Identify cycle members
+    const sccs = fkGraph.getSCCs();
+    for (const scc of sccs) {
+      if (scc.length > 1) {
+        for (const t of scc) cycleMembers.add(t);
+      }
+    }
+
+    // Add tables not in FK graph (no relationships)
+    for (const t of byTable.keys()) {
```
Comment (sanghyun-io) [inline src/scripts/migration-plan/execution-order.ts]: ### Issue #11 [MED]: O(n²) in computeExecutionOrder

`orderedTables.includes(t)` inside loop causes quadratic behavior.

**Suggested fix**: Use a `Set` for O(1) membership checks.

File: src/scripts/visualization/sugiyama-layout.ts [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+    for (const v of neighbors) {
+      const currentLayer = layer.get(v);
+      const newLayer = uLayer + 1;
+
+      if (currentLayer === undefined || newLayer > currentLayer) {
+        layer.set(v, newLayer);
+      }
+
+      inDegree.set(v, (inDegree.get(v) ?? 1) - 1);
+      if ((inDegree.get(v) ?? 0) <= 0 && !queue.includes(v)) {
```
Comment (sanghyun-io) [inline src/scripts/visualization/sugiyama-layout.ts]: ### Issue #12 [MED]: O(n²) queue membership in Sugiyama

`queue.includes(v)` inside traversal adds repeated linear scans.

**Suggested fix**: Maintain an `enqueued` Set alongside the queue for O(1) checks.

File: src/scripts/fix-options/strategies/charset-fix.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+      reversibility: 'partially_reversible',
+      rollbackTemplate: null,
+    });
+
+    // FK safe mode: convert in topological order
+    options.push({
+      strategy: 'collation_fk_safe',
+      label: 'FK 안전 순서로 변환',
+      description: `FK 의존성 순서를 따라 한 테이블씩 순서대로 변환합니다. 가장 안전하지만 느립니다.`,
+      sqlTemplate: [
```
Comment (sanghyun-io) [inline src/scripts/fix-options/strategies/charset-fix.ts:65]: ### Issue #1 [HIGH]: FK-safe charset fix is a no-op

`collation_fk_safe` option has a comment-only `sqlTemplate` with no executable ALTER. The recommender selects this for FK-related charset issues, so plans can "succeed" without applying fixes.

**Suggested fix**: Either generate real ordered ALTER TABLE statements from FK topo order, or set `sqlTemplate: null` and mark as manual-only.

File: src/scripts/analysis/charset-cascade.ts [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+  }
+  const col = findColumn(table, colName);
+  if (!col) {
+    return `MODIFY COLUMN \`${colName}\` /* 원본 타입 확인 필요 */ CHARACTER SET utf8mb4`;
+  }
+
+  const parts = [`MODIFY COLUMN \`${colName}\` ${col.type} CHARACTER SET utf8mb4`];
+  if (!col.nullable) parts.push('NOT NULL');
+  if (col.default !== undefined) {
+    parts.push(`DEFAULT ${col.default === 'NULL' ? 'NULL' : `'${col.default}'`}`);
```
Comment (sanghyun-io) [inline src/scripts/analysis/charset-cascade.ts]: ### Issue #2 [HIGH]: Generated charset fix SQL can be invalid for expression defaults

`buildColumnModify` blindly quotes defaults (`DEFAULT '${col.default}'`), which breaks expression defaults (`CURRENT_TIMESTAMP`), numeric defaults, and cases needing escaping.

**Suggested fix**: Detect expression vs literal defaults, emit unquoted for expressions and properly escape string literals.

File: src/scripts/migration-plan/plan-generator.ts [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+      order: order++,
+      title: '설정 파일 변경',
+      description: `my.cnf에서 제거된 시스템 변수 ${configIssues.issues.length}개를 제거/수정합니다.`,
+      type: 'action',
+      required: true,
+    });
+  }
+
+  // FK disable if needed
+  if (summary.needsFKDisable) {
```
Comment (sanghyun-io) [inline src/scripts/migration-plan/plan-generator.ts:145]: ### Issue #3 [MED]: FK-check lifecycle split across two layers

FK enable/disable appears both as a global plan step here AND inside charset fix SQL options, creating conflicting control flow.

**Suggested fix**: Centralize FK toggle policy in plan generation; strategy SQL should only contain DDL.

File: src/scripts/analysis/fk-graph.ts [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+        const parentKey = fk.refTable.toLowerCase();
+
+        // Check for self-reference
+        if (childKey === parentKey) {
+          this.selfRefTables.add(childKey);
+          // Include in graph but mark as self-ref
+        }
+
+        // Check for missing reference table
+        if (!this.allTables.has(parentKey)) {
```
Comment (sanghyun-io) [inline src/scripts/analysis/fk-graph.ts:78]: ### Issue #4 [MED]: missingReference context is effectively dead code

Missing parent edges are dropped during graph build, then missingReference checks parent sets that no longer include missing tables.

**Suggested fix**: Track missing parents via `getMissingTables()` API and populate missingReference from that.

File: src/scripts/analysis/fk-graph.ts [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+  /**
+   * BFS: Get all tables related to the given table (both directions).
+   */
+  getRelatedTables(table: string): Set<string> {
+    const key = table.toLowerCase();
+    const visited = new Set<string>();
+    const queue: string[] = [key];
+
+    while (queue.length > 0) {
+      const current = queue.shift()!;
```
Comment (sanghyun-io) [inline src/scripts/analysis/fk-graph.ts]: ### Issue #5 [MED]: queue.shift() O(n²) in BFS traversals

Core BFS traversals use `queue.shift()` in loops, which is O(n²) for large arrays. Combined with repeated linear lookups in charset analysis.

**Suggested fix**: Switch to index-based queue pattern (already used in sugiyama-layout.ts) and precompute per-table column maps.

File: src/scripts/analysis/fk-graph.ts [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+
+  /**
+   * Check whether the FK graph contains any cycles.
+   */
+  hasCycle(): boolean {
+    if (this.cachedHasCycle !== null) return this.cachedHasCycle;
+    const sccs = this.getSCCs();
+    this.cachedHasCycle = sccs.some(scc => scc.length > 1) ||
+      // A self-loop also counts as a cycle
+      this.selfRefTables.size > 0;
```
Comment (sanghyun-io) [inline src/scripts/analysis/fk-graph.ts:173]: ### Issue #8 [MED]: Self-referential cycles not propagated to planning

Self-reference counts as a cycle in graph logic, but cycle-member marking only uses SCCs with size > 1. Self-loop tables can be missed in hasCycle flags and FK-disable planning.

**Suggested fix**: Merge getSelfRefTables() results into cycle flags/summaries for consistent semantics.

**Review by sanghyun-io — COMMENTED** [review]
## Codex Review (2nd Round)

**Verdict: REVISE** — 8 issues found (2 HIGH, 6 MED)

| # | Severity | Area | Issue |
|---|:---:|:---:|---|
| 1 | **HIGH** | Impl | FK-safe charset fix is a no-op (comment-only SQL) |
| 2 | **HIGH** | Impl | charset fix SQL breaks expression defaults |
| 3 | MED | Arch | FK-check lifecycle split across two layers |
| 4 | MED | Impl | missingReference context is dead code |
| 5 | MED | Perf | queue.shift() O(n²) + linear lookups |
| 6 | MED | Test | Missing test for non-executable recommended SQL |
| 7 | MED | Test | Missing edge-case tests for default preservation |
| 8 | MED | Arch | Self-ref cycles not propagated to planning |

See inline comments for details.
